File: week16/태연/PG42579_베스트앨범.py
def solution(genres, plays):

    '''
    예전에 풀었던 파이썬 풀이인데 좋은풀이는아님
    근데 그냥 올려봄 

    테스트 1 〉	통과 (0.06ms, 10.4MB)
    테스트 2 〉	통과 (0.01ms, 10.2MB)
    테스트 3 〉	통과 (0.01ms, 10.4MB)
    테스트 4 〉	통과 (0.01ms, 10.1MB)
    테스트 5 〉	통과 (0.15ms, 10.3MB)
    테스트 6 〉	통과 (0.07ms, 10.2MB)
    테스트 7 〉	통과 (0.08ms, 10.3MB)
    테스트 8 〉	통과 (0.05ms, 10.1MB)
    테스트 9 〉	통과 (0.01ms, 10.2MB)
    테스트 10 〉통과 (0.08ms, 10.4MB)
    테스트 11 〉통과 (0.02ms, 10.2MB)
    테스트 12 〉통과 (0.05ms, 10.1MB)
    테스트 13 〉통과 (0.07ms, 10.2MB)
    테스트 14 〉통과 (0.08ms, 10.4MB)
    테스트 15 〉통과 (0.02ms, 10.1MB)
    '''
    
    answer = []
    song = []
    for i in range(len(genres)):
        song.append([genres[i],plays[i],i]) 
    song = sorted(song, key = lambda x:x[1],reverse=True)
    
    song_num = {}
    
    for s in song:
        if s[0] in song_num:
            song_num[s[0]] += s[1]
        else:
            song_num[s[0]] = s[1]
            
    song_num = sorted(song_num,key=lambda x:song_num[x], reverse=True)
    
    for kind in song_num:
        index = 0
        for s in song:
            if s[0] == kind and index < 2:
                answer.append(s[2])
                index +=1
        
    return answer

File: week16/태연/PG42579_베스트앨범.java
import java.util.*;

class Solution {

    /*
     *  - 시간 하단 첨부
     * 
     *  - TreeSet 여러개 사용
     */
    
    static class Songs implements Comparable<Songs>{
        TreeSet<Song> list;
        int plays;
        
        Songs(){
            this.list = new TreeSet<>();
            this.plays = 0;
        }
        
        void add(Song s){
            this.list.add(s);
            this.plays+=s.plays;
        }
        
        public int compareTo(Songs s){
            return s.plays-this.plays;
        }
    }
    
    static class Song implements Comparable<Song>{
        int id;
        int plays;
        
        Song(int id, int plays){
            this.id=id;
            this.plays=plays;
        }
        
        public int compareTo(Song s){
            if(this.plays==s.plays){
                return this.id-s.id;
            } else {
                return s.plays-this.plays;
            }
        }
    }
    
    public int[] solution(String[] genres, int[] plays) {
        
        HashMap<String,Integer> genreToIndex = new HashMap<>();
        List<Songs> songList = new ArrayList<>();
        int curIdx = 0;
        
        for(int i=0; i<genres.length; i++){
            String genre = genres[i];
            int play = plays[i];
            
            if(!genreToIndex.containsKey(genre)){
                genreToIndex.put(genre, curIdx++);
                songList.add(new Songs());
            }
            
            songList.get(genreToIndex.get(genre)).add(new Song(i, play));
        }
        
        Collections.sort(songList);
        
        Queue<Integer> q = new LinkedList<>();
        
        for(Songs songs : songList){
            q.add(songs.list.pollFirst().id);
            if(!songs.list.isEmpty()){
                q.add(songs.list.pollFirst().id);
            }
        }
        
        int[] ret = new int[q.size()];
        
        int idx=0;
        while(!q.isEmpty()){
            ret[idx++] = q.poll();
        }
        
        return ret;
    }
    
}

/*
     *  정확성  테스트
     * 테스트 1 〉	통과 (1.21ms, 76.2MB)
     * 테스트 2 〉	통과 (1.44ms, 74.9MB)
     * 테스트 3 〉	통과 (1.03ms, 75.4MB)
     * 테스트 4 〉	통과 (1.02ms, 82.7MB)
     * 테스트 5 〉	통과 (2.51ms, 72.9MB)
     * 테스트 6 〉	통과 (1.19ms, 71.6MB)
     * 테스트 7 〉	통과 (1.18ms, 72.8MB)
     * 테스트 8 〉	통과 (1.89ms, 78.2MB)
     * 테스트 9 〉	통과 (1.91ms, 79MB)
     * 테스트 10 〉	통과 (1.99ms, 83.1MB)
     * 테스트 11 〉	통과 (1.12ms, 81.1MB)
     * 테스트 12 〉	통과 (1.12ms, 74.3MB)
     * 테스트 13 〉	통과 (1.81ms, 77.5MB)
     * 테스트 14 〉	통과 (1.25ms, 78.6MB)
     * 테스트 15 〉	통과 (1.01ms, 71.2MB)
     */